File: DataStructurePython/binary_search.py
def binary_search(data: [int], key: int) -> int:
    lower_bound = 0
    upper_bound = len(data) - 1
    mid = 0
    while True:
        if lower_bound > upper_bound:
            return -1
        # 注意：直接使用 (lowerBound + upperBound) / 2 可能导致超出 int 范围溢出
        mid = lower_bound + (upper_bound - lower_bound) / 2
        if data[mid] == key:
            return mid
        elif data[mid] < key:
            lower_bound = mid + 1
        else:
            upper_bound = mid - 1

File: DataStructureSwift/Sources/DataStructureSwift/BinarySearch.swift
//
//  BinarySearch.swift
//  
//
//  Created by Yubo Qin on 11/29/22.
//

import Foundation

func binarySearch(data: [Int], key: Int) -> Int {
    guard !data.isEmpty else {
        return NSNotFound
    }

    var lowerBound = 0
    var upperBound = data.count - 1
    var mid = 0

    while true {
        if lowerBound > upperBound {
            return NSNotFound
        }
        // 注意：直接使用 (lowerBound + upperBound) / 2 可能导致超出 Int 范围溢出
        mid = lowerBound + (upperBound - lowerBound) / 2
        if data[mid] == key {
            return mid
        } else if data[mid] < key {
            lowerBound = mid + 1
        } else {
            upperBound = mid - 1
        }
    }
}
